docs/source/spec-workload-executor.rst

```diff
@@ -37,55 +37,137 @@ Behavioral Description
-      success or failure of a driver workload execution. A non-zero value for either of these fields is construed
-      as a sign that something went wrong while executing the workload and the test is marked as a failure.
-      The workload executor's exit code is **not** used for determining success/failure and is ignored.
-
-   .. note:: If ``astrolabe`` encounters an error in parsing the workload statistics dumped to ``results.json``
-      (caused, for example, by malformed JSON), ``numErrors``, ``numFailures``, and ``numSuccesses``
-      will be set to ``-1`` and the test run will be assumed to have failed.
-
-   .. note:: The choice of termination signal used by ``astrolabe`` varies by platform. ``SIGINT`` [#f1]_ is used as
-      the termination signal on Linux and OSX, while ``CTRL_BREAK_EVENT`` [#f2]_ is used on Windows.
-
-   .. note:: On Windows systems, the workload executor is invoked via Cygwin Bash.
+#. MUST parse the incoming ``driverWorkload`` document and set up
+    the driver's unified test runner to execute the provided workload.
+    
+    .. note::
+    
+      The workload SHOULD include a ``loop`` operation, as described in the
+      unified test format, but the workload executor SHOULD NOT validate that
+      this is the case.
+
+#. MUST set a signal handler for handling the termination signal that is
+   sent by ``astrolabe``. The termination signal is used by ``astrolabe``
+   to communicate to the workload executor, and ultimately the unified test
+   runner, that they should stop running operations.
+
+#. MUST invoke the unified test runner to execute the workload.
+   If the workload includes a ``loop`` operation, the workload will run until
+   terminated by the workload executor; otherwise, the workload will terminate
+   when the unified test runner finishes executing all of the operations.
+   The workload executor MUST handle the case of a non-looping workload and
+   it MUST terminate when the workload terminates.
+   
+   If the unified test runner raises an error while executing the workload,
+   the error MUST be reported using the same format as errors handled by the
+   unified test runner, as described in the unified test runner specification
+   under the ``loop`` operation. Errors handled by the workload
+   executor MUST be included in the calculated (and reported) error count.
+   
+   If the unified test runner reports a failure while executing the workload,
+   the failure MUST be reported using the same format as failures handled by the
+   unified test runner, as described in the unified test runner specification
+   under the ``loop`` operation. Failures handled by the workload
+   executor MUST be included in the calculated (and reported) failure count.
+   If the driver's unified test runner is intended to handle all failures
+   internally, failures that propagate out of the unified test runner MAY
+   be treated as errors by the workload executor.
```

Comment:
The original text clarified the difference between errors and failures, but this current text seems more vague. Is there any need to differentiate between errors and failures if some implementations may not be able to?

I realize the unified test format spec says things like "consider this test to have failed" and "MUST report an error", but I think there's a lot of variation among how that is implemented in drivers. In PHP's case, where we run the unified tests as part of our larger test suite, any unexpected error (e.g. "failure" due to operation assertion failing, "error" due to unsupported operation) will bubble up and be regarded as a test failure.

To take this a step further: will most workloads even need to use result/error assertions? I was under the impression that Atlas testing mainly wanted to execute some operations in a loop and rely on features like retryable writes to minimize/avoid errors (e.g. command failures, socket exceptions). Is there a use case for wanting to assert result objects like we do for other API-centric spec tests?

Reply:
I believe we discussed this is as part of the unified test format PR, but to provide context in this PR: the idea behind separating errors and failures is that an error generally means the test scenario author made an error in the scenario (e.g. misspelled an operation name), whereas a failure generally means the driver under test or Atlas failed to operate as expected (e.g. a write retry failed). Thus I believe there is value in distinguishing the two, although presently the reporting mechanism for them both looks exactly the same.

Reply:
I added a provision to the workload executor specification to set the errors or failures that are not reported by its unified test runner to the empty array.  Thus the scenarios should be written to collect both, and astrolabe should be able to assume that both would be present.